### PR TITLE
`Contributor Guides` - Add new rules into current guides

### PR DESCRIPTION
**Pull request by ms-zhenhua**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [ ] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [ ] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-merging.md).

* `azurerm_resource` - support for the `thing1` property [GH-00000]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)

Fixes #0000

## AI Assistance Disclosure

- [ ] AI Assisted - This contribution was made by, or with the assistance of, AI/LLMs

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the provider.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

**Review by ms-henglu — APPROVED** [review]
LGTM

**Comment by SC-Tobias** [comment]
Consideration should also be given to moving the [README.md](https://github.com/hashicorp/terraform-provider-azurerm/blob/3792f533b07fe1781962a4b370fbf4c94068ae6e/contributing/README.md) file to the root folder ("docs" or ".github" folder) and renaming it to CONTRIBUTING.md, as this will be displayed as a new tab on the Repo Startpage. Also a info appears if you open a pr.

https://github.com/hashicorp/terraform-provider-azurerm/blob/3792f533b07fe1781962a4b370fbf4c94068ae6e/contributing/README.md

For more infos see:

https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors#about-contributing-guidelines

File: contributing/topics/schema-design-considerations.md [hunk 1]
```diff
@@ -332,6 +336,8 @@ String arguments must be validated. Use `StringNotEmpty` at a minimum but ideall
 
 Numeric arguments should specify a valid range.
 
+Properties which are purely `Computed` (and not `Optional`) do not require a `ValidateFunc`.
```
Comment (wuxu92) [inline contributing/topics/schema-design-considerations.md:339]: tfproviderlint can check on this: https://github.com/bflad/tfproviderlint?tab=readme-ov-file#standard-schema-checks  we can have a look if the current CI enabled it rather than add it to the guidance.

File: contributing/topics/schema-design-considerations.md [hunk 2]
```diff
@@ -392,3 +398,18 @@ Numeric arguments should specify a valid range.
 	ValidateFunc: validation.IntBetween(32, 16384),
 },
 ```
+
+## Validating Enums
+
+When defining validation for a field where all Enumeration values defined in the SDK are valid options, use the built-in `PossibleValuesFor...` helper functions to define the allowed values instead of manually enumerating them. This approach reduces the risk of drift between the provider and the SDK.
```
Comment (wuxu92) [inline contributing/topics/schema-design-considerations.md]: The guide should also mention that all possible value should be confirmed or tested that are supported before we release in the provider, or we should explicitly expose these supported values

File: contributing/topics/guide-new-resource.md [hunk 3]
```diff
@@ -686,6 +686,21 @@ func (r ExampleResource) CustomizeDiff() sdk.ResourceFunc {
 }
 ```
 
+- When the return type of a `flatten...` function is a slice, it should return an empty slice instead of `nil` when the input is `nil` or empty. The `nil` check for the `input` parameter should be handled within the `flatten...` function. Avoid performing this check prior to calling the function.
```
Comment (wuxu92) [inline contributing/topics/guide-new-resource.md:694]: for "it should return an empty slice instead of `nil` when the input is `nil` or empty" this is actually not always required, although the existing code usually follow it, sdkv2 actually treat nil and empty slice as the same, and for framework which will be adapted in a not too far future, this is actually not true, framework will differ nil and  empty slice and we should not return empty slice for nil. so we may not put this in our guidance.

File: contributing/topics/reference-acceptance-testing.md [hunk 4]
```diff
@@ -74,6 +74,12 @@ func (ExampleResource) preCheck(t *testing.T) {
 
 Go does not require a specific function order, but for readability it is recommended to place `preCheck` close to the tests that call it (commonly after the `TestAcc...` functions and before the `Exists`/`Destroy` methods), following the pattern used throughout `internal/services/*/*_test.go`.
 
+### ComposeTestCheckFunc
+
+Do not validate fields using `check.That(...).Key(...).Exists()`. The testing framework inherently assists in verifying field changes, making this explicit check unnecessary.
```
Comment (wuxu92) [inline contributing/topics/reference-acceptance-testing.md]: need to mention that for computed or O+C attributes, this check is required if the acc test if to fix or validate the logic of the Computed or O+C attribute

File: contributing/topics/schema-design-considerations.md [hunk 5]
```diff
@@ -257,6 +257,10 @@ Would be better exposed as the following resources:
 - `azurerm_data_factory_linked_service_azure_function`
 - `azurerm_data_factory_linked_service_azure_search`
 
+## Common Fields Excluded from the Schema
+
+Some top-level common fields of Azure APIs should not be exposed in the schema, such as `type`, `state`, `provisioning_state`, etc.
```
Comment (wuxu92) [inline contributing/topics/schema-design-considerations.md]: we can make it clearer that some properties, rather than these "top-level" fields, which are not useful for our customer, like type, state, provisioning_state should not be exposed in Terraform schema.

File: contributing/topics/best-practices.md [hunk 6]
```diff
@@ -277,13 +277,46 @@ if _, err := client.Delete(ctx, newId, apirelease.DeleteOperationOptions{IfMatch
...
+When initializing a variable or struct field, avoid explicitly setting it to its zero value. In Go, variables are automatically initialized to their zero value, so setting them explicitly is redundant.
+
+:no_entry_sign: **DON'T**
+
+```go
+model.Profile = &profiles.Profile{
+	storage_enabled: false, // redundant: `false` is the default value
+	rules:           nil,   // redundant: `nil` is the default value
+	description:     "",    // redundant: `""` is the default value
+	count:           0,     // redundant: `0` is the default value
```
Comment (wuxu92) [inline contributing/topics/best-practices.md:397]: This is a kind of golang language preference rather the provider one, wondering if there is a golangci lint rule for it, if yes then we don't need it in the provider guidance best practice.

**Review by wuxu92 — COMMENTED** [review]
Thanks, these guidance rules are really useful! I left several comments otherwise LGTM

File: contributing/topics/best-practices.md (hunk 6 above)
Comment (ms-zhenhua) [reply contributing/topics/best-practices.md:397]: This rule comes from this [comment](https://github.com/hashicorp/terraform-provider-azurerm/pull/30115#discussion_r2231520797). AFAIK, there is no linter rule for checking this situation. I prefer to leave it to Hashi to confirm whether we need to add this new rule. WDYT?

File: contributing/topics/guide-new-resource.md (hunk 3 above)
Comment (ms-zhenhua) [reply contributing/topics/guide-new-resource.md:694]: This rule comes from this [comment](https://github.com/hashicorp/terraform-provider-azurerm/pull/30060#discussion_r2416426526). I prefer to leave it to Hashi to give us more suggestion that which style we should follow in future. WDYT?

File: contributing/topics/schema-design-considerations.md (hunk 1 above)
Comment (ms-zhenhua) [reply contributing/topics/schema-design-considerations.md:339]: I found [S010](https://github.com/bflad/tfproviderlint/blob/main/passes/S010) can check this issue. However, this rule comes from [this comment](https://github.com/hashicorp/terraform-provider-azurerm/pull/30086#discussion_r2192913101). I prefer to leave it to Hashi to help confirm whether we can enable this check in future pull requests. WDYT?

File: contributing/topics/schema-design-considerations.md (hunk 2 above)
Comment (ms-zhenhua) [reply contributing/topics/schema-design-considerations.md]: Updated. Thanks.

File: contributing/topics/reference-acceptance-testing.md (hunk 4 above)
Comment (ms-zhenhua) [reply contributing/topics/reference-acceptance-testing.md]: Updated with descriptions of checking values of fields. Thanks.

File: contributing/topics/schema-design-considerations.md (hunk 5 above)
Comment (ms-zhenhua) [reply contributing/topics/schema-design-considerations.md]: updated. Thanks.

**Comment by ms-zhenhua** [comment]
Hi @wuxu92 , thanks for the review. I have replied your comments and updated part of the rules. Could you please take another review?

File: contributing/topics/best-practices.md [hunk 7]
```diff
@@ -277,13 +277,46 @@ if _, err := client.Delete(ctx, newId, apirelease.DeleteOperationOptions{IfMatch
 }
 ```
 
-- Use `pointer.ToEnum` to convert Enum type instead of explicitly type conversion.
+- Use `pointer.ToEnum` to cast strings as an Enum type where API objects expect a pointer to the Enum value. For example: use `APIModel.SomeValue = pointer.ToEnum[someservice.SomeEnumType](model.SomeVariable)` instead of `APIModel.SomeValue = pointer.To(someservice.SomeEnumType(model.SomeVariable))`.
+
+- Use `pointer.FromEnum` to return a string from a pointer to an Enum without having to cast it. For example: use `myStruct.SomeStringValue = pointer.FromEnum(model.EnumValue)` instead of `myStruct.SomeStringValue = string(pointer.From(model.EnumValue))`.
```
Comment (WodansSon) [inline contributing/topics/best-practices.md:364]: Would it help if we added an actual code example so contributors would understand how to use it correctly?

```go
	if containerAppEnvironment.PublicNetworkAccess != "" {
		managedEnvironment.Properties.PublicNetworkAccess = pointer.ToEnum[managedenvironments.PublicNetworkAccess](containerAppEnvironment.PublicNetworkAccess)
	}
```

File: contributing/topics/best-practices.md [hunk 8]
```diff
@@ -277,13 +277,46 @@ if _, err := client.Delete(ctx, newId, apirelease.DeleteOperationOptions{IfMatch
...
+
+## Avoid Intermediate Variables
+
+Do not create a variable if it is used only once. For example, do not create a variable for the following `expandModelContent` and then assign it to `param.Content`, instead assign it directly.
 
 :white_check_mark: **DO**
```
Comment (WodansSon) [inline contributing/topics/best-practices.md]: These sort of emoji's ✅ are not generally used in the HashiCorp contributor guidelines, can we standardize these to follow the HashiCorp contributor guidelines style?

File: contributing/topics/best-practices.md [hunk 9]
```diff
@@ -277,13 +277,46 @@ if _, err := client.Delete(ctx, newId, apirelease.DeleteOperationOptions{IfMatch
...
-return &managedclusters.ManagedClusterBootstrapProfile{
-    ArtifactSource: pointer.ToEnum[managedclusters.ArtifactSource](config["artifact_source"].(string)),
+param.Content = expandModelContent(model.Content)
+```
+
+## Do Not Initialize Variables with Their Zero Value
+
+When initializing a variable or struct field, avoid explicitly setting it to its zero value. In Go, variables are automatically initialized to their zero value, so setting them explicitly is redundant.
+
+:no_entry_sign: **DON'T**
```
Comment (WodansSon) [inline contributing/topics/best-practices.md]: Can we change this from the `no_entry_sign` emoji to the `x` github emoji? Most HashiCorp contributor docs use GitHub emoji shortcodes + bold uppercase labels.

```suggestion
:x: **DON'T**
```

File: contributing/topics/guide-new-resource.md [hunk 10]
```diff
@@ -921,6 +938,10 @@ resource "azurerm_resource_group_example" "test" {
 }
 ```
 
+Things worth noting here:
+
+- The `name` argument for each resource definition should start with `acctest`, unless the API backend imposes naming restrictions preventing this.
```
Comment (WodansSon) [inline contributing/topics/guide-new-resource.md]: Should we align the wording here with the other contributor documentation?

```suggestion
- The `name` argument within each resource definition should be prefixed with `acctest-` (unless this is disallowed by the API naming restrictions), avoid overly complex naming, and ensure any naming restrictions are followed. e.g. `name = "acctest-server"`.
```

File: contributing/topics/reference-acceptance-testing.md [hunk 11]
```diff
@@ -74,6 +74,10 @@ func (ExampleResource) preCheck(t *testing.T) {
 
 Go does not require a specific function order, but for readability it is recommended to place `preCheck` close to the tests that call it (commonly after the `TestAcc...` functions and before the `Exists`/`Destroy` methods), following the pattern used throughout `internal/services/*/*_test.go`.
 
+### ComposeTestCheckFunc
+
+Do not validate fields using `check.That(...).Key(...).Exists()`. The testing framework already verifies field changes, making this explicit check unnecessary. If you need to validate the value of a field, use `check.That(...).Key(...).HasValue()`, `check.That(...).Key(...).IsEmpty()` or `check.That(...).Key(...).IsNotEmpty()`.
```
Comment (WodansSon) [inline contributing/topics/reference-acceptance-testing.md]: For most resources this is solid advice, however should this be more descriptive and explain the edge cases too? This will only be automatically validated if you add the `data.ImportStep(),` step to the test case, also Data Sources currently still require the check step. There are also special cases where you should still use these checks to verify that the O+C/ForceNew fields are being validated correctly, for example if you have an O+C field and you remove it from the config, there should be a test that verifies the correct behavior for the API is being followed, in which case the `check.That` would be needed right?

File: contributing/topics/schema-design-considerations.md (hunk 1 above)
Comment (WodansSon) [reply contributing/topics/schema-design-considerations.md:339]: I agree with @ms-zhenhua, while this might be an automatic check it is good to document this so new contributors can understand this without having to get a lint error.

**Review by WodansSon — CHANGES_REQUESTED** [review]
@ms-zhenhua, thanks for this PR, this looks like a solid update to the docs. I did leave a few comments and questions. If you can give those a look at your earliest convenience I would be happy to give it another look! 🚀  

Current State: Requested Changes

File: contributing/topics/best-practices.md (hunk 7 above)
Comment (ms-zhenhua) [reply contributing/topics/best-practices.md:364]: added

File: contributing/topics/best-practices.md (hunk 9 above)
Comment (ms-zhenhua) [reply contributing/topics/best-practices.md]: updated

File: contributing/topics/guide-new-resource.md (hunk 10 above)
Comment (ms-zhenhua) [reply contributing/topics/guide-new-resource.md]: updated

File: contributing/topics/reference-acceptance-testing.md (hunk 11 above)
Comment (ms-zhenhua) [reply contributing/topics/reference-acceptance-testing.md]: updated

**Comment by ms-zhenhua** [comment]
Hi @WodansSon , thanks for your review and I have updated the documents. Could you kindly take another look?

**Review by WodansSon — DISMISSED** [review]
@ms-zhenhua, thank you for pushing those changes, this LGTM! 🚀

**Review by wuxu92 — APPROVED** [review]
Thanks, LGTM!

File: contributing/topics/schema-design-considerations.md (hunk 1 above)
Comment (wuxu92) [reply contributing/topics/schema-design-considerations.md:339]: I don't want the contributor document to include everything: especially items that can be detected by a linter. A short, compact guide is easier for new contributors to read and understand. While including everything would make an AI agent easier to follow, the documentation should remain concise.

File: contributing/topics/best-practices.md [hunk 12]
```diff
@@ -359,7 +359,21 @@ if _, err := client.Delete(ctx, newId, apirelease.DeleteOperationOptions{IfMatch
 }
 ```
 
-- Use `pointer.ToEnum` to convert Enum type instead of explicitly type conversion.
+- Use `pointer.ToEnum` to cast strings as an Enum type where API objects expect a pointer to the Enum value. For example: use `APIModel.SomeValue = pointer.ToEnum[someservice.SomeEnumType](model.SomeVariable)` instead of `APIModel.SomeValue = pointer.To(someservice.SomeEnumType(model.SomeVariable))`.
```
Comment (sreallymatt) [inline contributing/topics/best-practices.md]: Now covered by a linter: https://github.com/katbyte/azproviderlint/tree/main/checks/AZG/AZG003_pointer_to_enum_conversion

File: contributing/topics/best-practices.md [hunk 13]
```diff
@@ -359,7 +359,21 @@ if _, err := client.Delete(ctx, newId, apirelease.DeleteOperationOptions{IfMatch
...
+
+- Use `pointer.FromEnum` to return a string from a pointer to an Enum without having to cast it. For example: use `myStruct.SomeStringValue = pointer.FromEnum(model.EnumValue)` instead of `myStruct.SomeStringValue = string(pointer.From(model.EnumValue))`.
```
Comment (sreallymatt) [inline contributing/topics/best-practices.md]: Not yet covered by a linter but instead of documenting this we'll implement one

File: contributing/topics/best-practices.md [hunk 14]
```diff
@@ -359,7 +359,21 @@ if _, err := client.Delete(ctx, newId, apirelease.DeleteOperationOptions{IfMatch
...
+
+:white_check_mark: **DO**
+
+```go
+if containerAppEnvironment.PublicNetworkAccess != "" {
+	managedEnvironment.Properties.PublicNetworkAccess = pointer.ToEnum[managedenvironments.PublicNetworkAccess](containerAppEnvironment.PublicNetworkAccess)
+}
+```
+
+## Avoid Intermediate Variables
```
Comment (sreallymatt) [inline contributing/topics/best-practices.md]: Now covered by linter: https://github.com/katbyte/azproviderlint/tree/main/checks/AZG/AZG005_single_use_temporary

File: contributing/topics/best-practices.md [hunk 15]
```diff
@@ -368,3 +382,33 @@ return &managedclusters.ManagedClusterBootstrapProfile{
     ArtifactSource: pointer.ToEnum[managedclusters.ArtifactSource](config["artifact_source"].(string)),
 }
 ```
+
+## Do Not Initialize Variables with Their Zero Value
```
Comment (sreallymatt) [inline contributing/topics/best-practices.md]: Now largely covered by linter: https://github.com/katbyte/azproviderlint/tree/main/checks/AZG/AZG007_redundant_zero_value_field

maps and slices set to `nil` are not yet reported, but I will see about modifying that

File: contributing/topics/best-practices.md [hunk 16]
```diff
@@ -368,3 +382,33 @@ return &managedclusters.ManagedClusterBootstrapProfile{
...
+```go
+model.Profile = &profiles.Profile{
+	storage_enabled: false, // redundant: `false` is the default value
+	rules:           nil,   // redundant: `nil` is the default value
+	description:     "",    // redundant: `""` is the default value
+	count:           0,     // redundant: `0` is the default value
+}
+```
+
+## Slice and Map Initialization
```
Comment (sreallymatt) [inline contributing/topics/best-practices.md]: We should just write a linter for this, added to my TODOs

File: contributing/topics/guide-new-resource.md [hunk 17]
```diff
@@ -691,6 +691,36 @@ func (r ExampleResource) CustomizeDiff() sdk.ResourceFunc {
 }
 ```
 
+- When the return type of a `flatten...` function is a slice, it should return an empty slice instead of `nil` when the input is `nil` or empty. The `nil` check for the `input` parameter should be handled within the `flatten...` function. Avoid performing this check prior to calling the function.
```
Comment (sreallymatt) [inline contributing/topics/guide-new-resource.md]: This is covered by this linter: https://github.com/katbyte/azproviderlint/tree/main/checks/AZR/AZR008_flatten_returns_nil_slice

File: contributing/topics/schema-design-considerations.md (hunk 1 above)
Comment (sreallymatt) [reply contributing/topics/schema-design-considerations.md]: I agree with @wuxu92 here, we should lint where possible and ensure contributing guides stay concise. More text is not always a good thing (both for humans and LLMs)

File: contributing/topics/schema-design-considerations.md [hunk 18]
```diff
@@ -392,3 +398,18 @@ Numeric arguments should specify a valid range.
 	ValidateFunc: validation.IntBetween(32, 16384),
 },
 ```
+
+## Validating Enums
```
Comment (sreallymatt) [inline contributing/topics/schema-design-considerations.md]: There is now a linter for this: https://github.com/katbyte/azproviderlint/tree/main/checks/AZS/AZS004_enum_validation_possible_values
